src/pip/_internal/utils/unstable.py

```diff
@@ -0,0 +1,26 @@
+
+from pip._internal.exceptions import UnknownUnstableFeatures
+
+
+class UnstableFeaturesHelper(object):
+    """Handles logic for registering/validating/checking
+    """
+
+    def __init__(self):
+        super(UnstableFeaturesHelper, self).__init__()
+        self._names = set()
+        self._enabled_names = set()
+
+    def register(self, *names):
+        self._names.update(set(names))
+
+    def validate(self, given_names):
```

Comment:
It seems like it would be better to add the validation logic to the option parser itself, if possible. The optparse docs give an example of how to do this here: https://docs.python.org/3.7/library/optparse.html#adding-new-types
E.g. then you don't have to make a separate call to `validate()` inside `main()`, etc.

Also, then you can add a test that `parse_args()` raises the error you want. Currently, even though you're testing `validate()`, it doesn't look like you're testing that `validate()` is actually called inside `main()`.

Reply:
I don't want to dwell too deep into `optparse` tbh. I'd actually looked into that specific example but the fact that we'd need to use a custom Option subclass felt a little unnecessary. Further, keeping this separate makes it easier to test and it's overall a simpler change.

> it doesn't look like you're testing that validate() is actually called inside main().

Indeed. I'll look into adding a test for this. :)